File: expired-sale-prices-never-revert/node/revert-expired-sales.js
/**
 * Revert WooCommerce sale prices whose sale window already ended.
 *
 * A scheduled sale finishes on paper, but the wc_scheduled_sales WP-Cron task
 * that should clear it never runs (WP-Cron disabled, no overnight traffic, a
 * migration, a plugin conflict). This walks every product WooCommerce
 * currently flags as on sale, compares its stored sale end date to now, and
 * clears the sale price (and sale dates) for any product whose sale window
 * has passed. Regular price is never touched. Safe by default. Run on a
 * schedule.
 *
 * Guide: https://www.allanninal.dev/woocommerce/expired-sale-prices-never-revert/
 */
import { pathToFileURL } from "node:url";

const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

/** Parse a WooCommerce *_gmt date string (naive, UTC) into a Date. */
export function parseGmt(value) {
  if (!value) return null;
  return new Date(value.endsWith("Z") ? value : `${value}Z`);
}

/** Pull the fields that matter for the decision out of a product payload. */
export function saleWindowOf(product) {
  return {
    salePrice: product.sale_price || "",
    regularPrice: product.regular_price || "",
    endsAt: parseGmt(product.date_on_sale_to_gmt),
  };
}

/**
 * Pure decision: should this product's sale price be reverted right now?
 * No I/O here on purpose, so this can be unit tested without a live store.
 */
export function decide(saleWindow, now) {
  if (!saleWindow.salePrice) return ["skip", "no sale price set"];
  if (!saleWindow.endsAt) return ["skip", "open-ended sale, no end date"];
  if (saleWindow.endsAt > now) return ["skip", "sale window still open"];
  return ["revert", "sale end date has passed"];
}

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  if (!res.ok) throw new Error(`Woo ${path} returned ${res.status}`);
  return res.json();
}

async function* productsOnSale() {
  let page = 1;
  while (true) {
    const batch = await woo(`/products?on_sale=true&per_page=50&page=${page}`);
    if (!batch.length) return;
    for (const product of batch) yield product;
    page++;
  }
}

async function revertSale(productId) {
  await woo(`/products/${productId}`, {
    method: "PUT",
    body: JSON.stringify({ sale_price: "", date_on_sale_from: null, date_on_sale_to: null }),
  });
}

export async function run() {
  const now = new Date();
  let reverted = 0;
  for await (const product of productsOnSale()) {
    const window = saleWindowOf(product);
    const [action, reason] = decide(window, now);
    if (action !== "revert") continue;
    console.log(`Product ${product.id}: ${reason}. ${DRY_RUN ? "would revert" : "reverting"}`);
    if (!DRY_RUN) await revertSale(product.id);
    reverted++;
  }
  console.log(`Done. ${reverted} product(s) ${DRY_RUN ? "to revert" : "reverted"}.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((e) => { console.error(e); process.exit(1); });
}

File: expired-sale-prices-never-revert/python/revert_expired_sales.py
"""Revert WooCommerce sale prices whose sale window already ended.

An order can end up looking fine while the catalog quietly stays wrong: a
scheduled sale finishes on paper, but the wc_scheduled_sales WP-Cron task that
should clear it never runs (WP-Cron disabled, no overnight traffic, a
migration, a plugin conflict). This walks every product WooCommerce currently
flags as on sale, compares its stored sale end date to now, and clears the
sale price (and turns off the sale dates) for any product whose sale window
has passed. Regular price is never touched. Safe by default. Run on a
schedule.

Guide: https://www.allanninal.dev/woocommerce/expired-sale-prices-never-revert/
"""
import os
import logging
from datetime import datetime, timezone
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("revert_expired_sales")

WOO_URL = os.environ.get("WOO_STORE_URL", "https://example.com").rstrip("/")
AUTH = HTTPBasicAuth(
    os.environ.get("WOO_CONSUMER_KEY", "ck_dummy"),
    os.environ.get("WOO_CONSUMER_SECRET", "cs_dummy"),
)
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def parse_gmt(value):
    """Parse a WooCommerce *_gmt date string (naive, UTC) into an aware datetime."""
    if not value:
        return None
    return datetime.fromisoformat(value).replace(tzinfo=timezone.utc)


def sale_window_of(product):
    """Pull the fields that matter for the decision out of a product payload."""
    return {
        "sale_price": product.get("sale_price") or "",
        "regular_price": product.get("regular_price") or "",
        "ends_at": parse_gmt(product.get("date_on_sale_to_gmt")),
    }


def decide(sale_window, now):
    """Pure decision: should this product's sale price be reverted right now?

    No I/O here on purpose, so this can be unit tested without a live store.
    """
    if not sale_window["sale_price"]:
        return ("skip", "no sale price set")
    if sale_window["ends_at"] is None:
        return ("skip", "open-ended sale, no end date")
    if sale_window["ends_at"] > now:
        return ("skip", "sale window still open")
    return ("revert", "sale end date has passed")


def products_on_sale():
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/products",
            params={"on_sale": "true", "per_page": 50, "page": page},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for product in batch:
            yield product
        page += 1


def revert_sale(product_id):
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/products/{product_id}",
        json={"sale_price": "", "date_on_sale_from": None, "date_on_sale_to": None},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    now = datetime.now(timezone.utc)
    reverted = 0
    for product in products_on_sale():
        window = sale_window_of(product)
        action, reason = decide(window, now)
        if action != "revert":
            continue
        log.info(
            "Product %s: %s. %s",
            product["id"], reason, "would revert" if DRY_RUN else "reverting",
        )
        if not DRY_RUN:
            revert_sale(product["id"])
        reverted += 1
    log.info("Done. %d product(s) %s.", reverted, "to revert" if DRY_RUN else "reverted")


if __name__ == "__main__":
    run()
